sklearn/utils/validation.py

```diff
@@ -239,19 +240,30 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
     spmatrix_converted : scipy sparse matrix.
         Matrix that is ensured to have an allowed type.
     """
-    if accept_sparse in [None, False]:
-        raise TypeError('A sparse matrix was passed, but dense '
-                        'data is required. Use X.toarray() to '
-                        'convert to a dense numpy array.')
     if dtype is None:
         dtype = spmatrix.dtype
 
     changed_format = False
-    if (isinstance(accept_sparse, (list, tuple))
-            and spmatrix.format not in accept_sparse):
-        # create new with correct sparse
-        spmatrix = spmatrix.asformat(accept_sparse[0])
-        changed_format = True
+
+    if isinstance(accept_sparse, str):
+        accept_sparse = [accept_sparse]
+
+    if accept_sparse in [None, False]:
+        raise TypeError('A sparse matrix was passed, but dense '
+                        'data is required. Use X.toarray() to '
+                        'convert to a dense numpy array.')
+    elif (isinstance(accept_sparse, (list, tuple)) and
+          len(accept_sparse) and
```

Comment:
Hm right the empty list would fail with the indexing, sorry I didn't think about that. But the ``not a sparse format`` error you get from the string seems better than the generic ``accept_parse`` error message below, so I'd remove that check.